docs/src/coding-guidelines.md

```diff
@@ -0,0 +1,130 @@
+
+# Code Denesting (Inversion)
+
+Strive to minimize unnecessary indentation without compromising code clarity. One effective way to achieve this is by using early exits and the continue keyword in long loops.
+
+Consider following example:
+```c++
+// ...
+if (modified && isDpsOp(op))
+{
+    // ... some long code ...
+}
+```
+
+It is strongly recommended to format the code as follows:
+```c++
+// ...
+// We just need to transform the op if it was modified
+if (!modified) return;
+
+// We also just need to transform the op if it is DPS op
+if (!isDpsOp(op)) return;
+
+// .. some long code ...
+```
+
+This reduces loop nesting, makes the reasoning behind the conditions clearer, and signals to the reader that there is no subsequent else to worry about, reducing cognitive load. This can significantly improve code readability and comprehension.
+
+# Using namespaces
+
+Namespaces are an important part of C++ programming, providing a way to organize code and avoid naming conflicts. Choose namespace names that reflect the purpose or functionality of the code contained within.
+
+Follow these guidelines when defining namespaces:
+    - Use *lower-case* letters for short, single-word names or those with a clear acronym (e.g., ttnn, mlir).
+    - Use nested namespaces to group logically related code, avoiding too deep or unnecessarily complex hierarchy
+    - Avoid placing code in the global namespace to reduce the potential for name conflicts and ambiguity. Always use specific namespaces. If necessary to use something from the global namespace (such as std), use an explicit std:: prefix rather than importing everything using using namespace std;
+    - Never use using namespace in header files, as it can lead to namespace pollution and potential conflicts. If you must use it, do so in .cpp (source) files where the scope is limited. Instead of importing entire namespaces, use specific imports to avoid ambiguity.
+```c++
+using std::vector;
+using std::string;
+```
+    - When namespaces are long or difficult to type repeatedly, use aliases for ease of use, but keep them meaningful.
+```c++
+namespace utils = mlir::tt::ttnn::utils;
+```
+    - Try to avoid mixing concepts from different namespaces in a single function or class. If a function belongs to one namespace but calls classes from others, ensure the relationships are clear.
+
+# Using Alternative Tokens (and, or, xor, etc.)
+
+Although they are standard, we should avoiding their use. They are very rarely used in practice and the C++ community widely uses the standard operators (&&, ||, !, etc.), as they are more familiar and easily recognizable to most C++ developers. Their usage can make the code harder to read and maintain, especially for developers who are not familiar with these alternatives. We should stick to the standard operators (&&, ||, !, etc.) for clarity, consistency, and compatibility with other C++ developers and tools.
```

Comment:
my vote would be against this guideline choice: the alternative tokens can be more readable because clang-format's choice is to omit some white space where it would actually make sense (e.g. the "!" is easy to miss in "!condition" vs "not condition"). I am observing this syntax being more widely used in template/c++-20 concept conditioning.

Reply:
The whole codebase is in "traditional" tokens. It would probably be a mess for a while if we started switching at this point, not sure it's worth the headache.

Reply:
I am ok keeping with the existing project style.

However, the language has other syntax variants that have identical or subsuming semantics, e.g. type aliasing via `typedef` or `using`. Perhaps make a similar guideline that `typedef` should be avoided in preference to `using`?

Reply:
I would keep the existing project style as it is already used in LLVM ecosystem. Regarding the other things, let me check what LLVM prefers in that regard, and let's stick to that.

Reply:
;)
```
        while (mask.any() and not mask[i]) {
          ++i;
        }
```

Reply:
Yeah, let's avoid typedef in preference for using.